### ISSUE TITLE
Jupyter output is directed to console, not browser

### ISSUE THREAD

**Issue by esafak** [scopatz/xonsh#629]
It is counter-intuitive and atypical of Jupyter kernels for the output of a command to be sent only to the console, as it currently seems to. If this is due to misconfiguration, please change the defaults so that this does not occur, and amend the documentation to explain how to change this behavior.
 
This applies to the [latest git commit](https://github.com/scopatz/xonsh/commit/b26568ede680f9869e2de53800478ddb295e87c6) installed with pip3 on Ubuntu 14.04.3.

**Comment by scopatz** [comment]
Thanks for reporting @esafak. This definitely needs fixing.

**Comment by blink1073** [comment]
I has just about to report this, as I am trying to update the metakernel version to use your recent changes.  I'll see if I can figure anything out.

**Comment by blink1073** [comment]
We're using `sys.stdout.write` in metakernel and that works, it appears you are doing the same here: https://github.com/scopatz/xonsh/blob/master/xonsh/tools.py#L311

**Comment by blink1073** [comment]
One problem is that only the second context manager was called, it should be:

```python
with redirect_stdout(out):
    with redirect_stderr(err):
        shell.default(code)
```

The other problem seems to be that the stdout is not affecting the context of `shell.default`, still digging.

**Comment by blink1073** [comment]
I tinkered with the options passed to `compile` and `exec`, but I couldn't figure out a way to capture the `stdout`....

**Comment by scopatz** [comment]
Hi @blink1073, we do a lot of redirection, which is probably what is confusing everything.  Out of curiosity, do you have `$XONSH_STORE_STDOUT = True` in your xonshrc. I wonder if that is causing additional issues.

**Comment by scopatz** [comment]
For what it is worth @blink1073, you can check out #634 for what I had to do to get redirect working. Redirecting the uncaptured subprocesses is the trickiest. I ended up implementing new builtin variables that you can attach a file-like object to.

**Comment by blink1073** [comment]
I can confirm #634 fixes the issue.  I tried it with and without `$XONSH_STORE_STDOUT = True` before  using #634 and it had no effect.

**Comment by blink1073** [comment]
I should be able to get this working on Windows.  If so, do you want to just include the Xonsh Metakernel as a dependency?

**Comment by blink1073** [comment]
I can give you commit rights to the repo.

**Comment by scopatz** [comment]
> If so, do you want to just include the Xonsh Metakernel as a dependency?

As a dependency for xonsh itself?

**Comment by blink1073** [comment]
Either that or not include it in `xonsh`.  I'd prefer to only have one.  If you want to keep yours, I'd deprecate mine.

**Comment by scopatz** [comment]
Personally, I'd rather not add more optional dependencies to xonsh itself and I am happy to have the version in their as it stands now.  I thought that the metakernel version did something slightly different, though, by spinning up a separate shell process and communicating via stdin / stdout / stderr.  The one in xonsh itself has exec's code more directly.  Am I mistaken about that and are these two truly redundant? I have glanced at your code but I don't really know how metakernel works....

**Comment by blink1073** [comment]
I was spinning up a subprocess because I ran into problems when trying to use it directly that you have subsequently solved.  I am about to roll in the changes you have made.  They will be functionally equivalent, with the addition of the magics that metakernel supports out of the box.

**Comment by blink1073** [comment]
There is also the problem of `pip` forcibly converting tar files to wheels, which will prevent your kernel from being installed by pip.

**Comment by scopatz** [comment]
Uggh really. I hate pip.  I am pretty sure that conda handles this just fine.  

So I guess my take on this is that - even with the direct usage - the kernels actually have slightly different use cases.  The one in xonsh is for if you just want something that works (but may not install) and yours has the added metakernel fanciness (installs, but may has an additional dependency).

We should probably document this distinction somewhere :)

**Comment by blink1073** [comment]
The problem is, the user ends up with two kernels.  I could call mine something different, but it is just too much confusion.

**Comment by blink1073** [comment]
Unless we remain in concert and mine overrides yours, I suppose that isn't *too* bad.

**Comment by scopatz** [comment]
Maybe "Calypso Xonsh"?  Or they can override... Either way :)

**Comment by blink1073** [comment]
okay, fair enough.

**Comment by scopatz** [comment]
This has been fixed by #634. Please reopen if you feel otherwise.
